test/realm_test.dart

```diff
@@ -829,6 +854,219 @@ Future<void> main([List<String>? args]) async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    await realm.writeAsync(() {
+      realm.add(Person('John'));
+    });
+
+    expect(realm.all<Person>().length, 1);
+  });
+
+  test('Realm.beginWriteAsync when realm is closed undoes changes', () async {
+    final realm1 = getRealm(Configuration.local([Person.schema]));
+    final realm2 = getRealm(Configuration.local([Person.schema]));
+
+    await realm1.beginWriteAsync();
+    realm1.add(Person('John'));
+    realm1.close();
+
+    expect(realm2.all<Person>().length, 0);
+  });
+
+  test("Realm.writeAsync with multiple transactions doesnt't deadlock", () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final t1 = await realm.beginWriteAsync();
+    realm.add(Person('Marco'));
+
+    final writeFuture = realm.writeAsync(() {
+      realm.add(Person('Giovanni'));
+    });
+
+    await t1.commitAsync();
+    await writeFuture;
+
+    final people = realm.all<Person>();
+    expect(people.length, 2);
+    expect(people[0].name, 'Marco');
+    expect(people[1].name, 'Giovanni');
+  });
+
+  test('Realm.writeAsync returns valid objects', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final person = await realm.writeAsync(() {
+      return realm.add(Person('John'));
+    });
+
+    expect(person.name, 'John');
+  });
+
+  test('Realm.writeAsync throws user exception', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    try {
+      await realm.writeAsync(() {
+        throw Exception('User exception');
+      });
+    } on Exception catch (e) {
+      expect(e.toString(), 'Exception: User exception');
+    }
+  });
+
+  test('Realm.writeAsync FIFO order ensured', () async {
+    final acquisitionOrder = <int>[];
+    final futures = <Future<void>>[];
+
+    final realm = getRealm(Configuration.local([Person.schema]));
+
+    for (var i = 0; i < 5; i++) {
+      futures.add(realm.writeAsync(() {
+        acquisitionOrder.add(i);
+      }));
+    }
+
+    await Future.wait(futures);
+
+    expect(acquisitionOrder, [0, 1, 2, 3, 4]);
+  });
+
+  test('Realm.beginWriteAsync with cancellation token', () async {
+    final realm1 = getRealm(Configuration.local([Person.schema]));
+    final realm2 = getRealm(Configuration.local([Person.schema]));
+    final t1 = realm1.beginWrite();
+
+    final token = TimeoutCancellationToken(const Duration(milliseconds: 1), timeoutException: CancelledException());
+
+    await expectLater(realm2.beginWriteAsync(token), throwsA(isA<CancelledException>()));
+
+    t1.rollback();
+  });
+
+  test('Realm.writeAsync with cancellation token', () async {
+    final realm1 = getRealm(Configuration.local([Person.schema]));
+    final realm2 = getRealm(Configuration.local([Person.schema]));
+    final t1 = realm1.beginWrite();
+
+    final token = CancellationToken();
+    Future<void>.delayed(Duration(milliseconds: 1)).then((value) => token.cancel());
```

Comment:
Is this testing it right. I suspect this is a race which is won by the Future so the token is canceled before the `writeAsync` has even started each time. 
Shouldn't we somehow trick it so we are sure the writeAsync had a chance to open a transaction and then the cancel happens.

Reply:
ie we are never calling `realm_async_begin_write` in this test

Reply:
In other words the delayed future here does not have any effect. 
In fact if I disable the `cancel` call here the test never completes with an failure.

Reply:
In fact if I comment out the `cancel` operation, this test never completes with a failure.

Reply:
Actually I think our use of `expecLater` is incorrect. This function should be used when we have async matchers,  and we don't have such. So using it here and awaiting on it is causing the test to hang. I pretty sure we use it incorrectly in most of the other places as well. But we can discuss that on the round table.

Reply:
`writeAsync` can't open a transaction because `t1` is open and write transactions are exclusive. I don't see a reason to believe there's a race here - `Future.delayed` continuation will execute when the thread is suspended, so it doesn't matter if it's 1 millisecond or 1 microseconds - it will always execute after whatever synchronous work is going on. The call to `writeAsync` is synchronous up until we return the `completer.future` and start waiting on Core.

I'm not sure what's wrong with our use of `expectLater` - the way I read the docs, it will execute the matchers when the future resolves. If you comment out the cancel you end up in a deadlock so it's not surprising the test hangs. The issue is that you're holding the write lock with `t1`, then awaiting the result of acquiring a write lock, which can not succeed until `t1` is relinquished. The fact that the test doesn't deadlock is evidence that the cancelation logic works.

Reply:
Resolved offline - for context - we added print statements to verify that the Core cancellation method is indeed called and the cancellation token is not already canceled by the time we reach `writeAsync`.

Reply:
test is working fine after an investigation